### ISSUE TITLE
Not correct description of -r|--required-parameter for complete builtin

### ISSUE THREAD

**Issue by EmilySeville7cfg** [fish-shell/fish-shell#8459]
Fish version: 3.3.1-557-gb550b3885 (built from source)
OS version: Linux Mint 20.2 Cinnamon

Now in the first case `-x` doesn't require any argument delimited by space from it while `-y` allows (but not requires) argument to be specified via space or to be joined with it:

```
complete --command some --short-option x --arguments='1 2'
complete --command some --short-option y --required-parameter --arguments='3 4'
```

`man complete` describes `-r` as:

> -r or --require-parameter says that this completion **must** have an option argument, i.e. may not be followed by  another option

But it's only almost correct: this option doesn't mandate user to specify argument alone, but simple permit it. There are two ways to fix it: change `-r`|`--required-parameter` behaviour or modify docs. ;)

**Comment by krobelus** [comment]
The docs are right, the implementation is probably too permissive. 
I guess you mean this special case that contradicts the "may not be followed by another option" bit:
Given `complete : -s y -ra "3 4"`, on `: -y -` we offer `-y` even though it's probably not meaningful.
We could not offer options in that case, this will only break users who used `-r` or `-x` in places where the argument is not actually requred. In practice no one has complained about that because options are only completed when there is a `-`.

**Comment by EmilySeville7cfg** [comment]
> I guess you mean this special case that contradicts the "may not be followed by another option" bit:

No, I am talking about option value (3 or 4) that can be joined or not (dependently on user) with `-y` (in our case). Documentation states that we can only use `-y 3` or `-y 3` but not `-y3` or `-y4`. Maybe better to edit docs as:

> -r or --require-parameter says that this completion **may** have an option argument, i.e. may not be followed by another option

**Comment by faho** [comment]
No, the option must have an argument. It's just that the argument can both be directly attached or given as another commandline argument.

**Comment by EmilySeville7cfg** [comment]
Suppose my completions are:

```fish
complete --command exe --erase
complete --command exe --short-option x --arguments='1 2'
complete --command exe --short-option y --require-parameter --arguments='3 4'
```

- When I enter `./exe -x ` and `tab` I expect `-x` arguments to be shown.
- When I enter `./exe -x -` and `tab` I expect no options shown because `-y` requires an argument value.

**Comment by krobelus** [comment]
> When I enter ./exe -x<space> and tab I expect 1 2 arguments to be shown but actually Fish shows only files and directories.

That's expected because the argument is optional; adding a space means the GNU-style option has no argument.

> When I enter ./exe -y - and tab I expect no options shown because -y requires an argument value but actually Fish shows -x -y.

yes, we should probably fix that

**Comment by EmilySeville7cfg** [comment]
> That's expected because the argument is optional; adding a space means the GNU-style option has no argument.

It's confusing because for short options arguments can be sticked to them or not: `-x1`, `-x 1` according to [docs](https://github.com/fish-shell/fish-shell/blob/4657ac0a7c05cc5da2b347b4c75eb434a30c6724/doc_src/cmds/complete.rst). ;) Please correct me if I am mistaken but now I think that `--required-parameter` doesn't affect short options because when there is no `--required-parameter` short options anyway can have argument attached to or detached from them. ;)
